### ISSUE TITLE
Internal error in `lift-singleton-groups` when having a datomic db value in fn args

### ISSUE THREAD

**Issue by avocade** [bhb/expound#205]
We've found an issue where `lift-singleton-groups` crashes when a datomic DB is one of the arguments. The crash is related to its use of `walk/postwalk`, which assumes DB can be treated as a regular map.

The problem can be reproduced by running:

```clojure
(require '[datomic.client.api :as d])
(walk/postwalk identity [(d/db conn)])
```

**Comment by eneroth** [comment]
The specific error message:

```clojure 
java.lang.ClassCastException: class datomic.core.db.Datum cannot be cast to class java.lang.Number (datomic.core.db.Datum is in unnamed module of loader 'app'; java.lang.Number is in module java.base of loader 'bootstrap')
```

**Comment by Jaretbinford** [comment]
A Datomic DB `(d/db conn)` is not a map and does not return a map.  It is a database.

**Comment by bhb** [comment]
I’m afraid I don’t fully understand (I don’t use Datomic, so I may be missing something).

Are you writing a spec for your database? What happens if you try to check the spec without using Expound e.g. using `s/valid?` or `s/explain` using spec?

**Comment by eneroth** [comment]
![Screenshot 2020-11-12 at 17 48 03](https://user-images.githubusercontent.com/333974/98969541-3b073f80-250f-11eb-9929-942c7594004a.png)

It implements the map protocol, which causes `postwalk` to blow up.

**Comment by eneroth** [comment]
> I’m afraid I don’t fully understand (I don’t use Datomic, so I may be missing something).
> 
> Are you writing a spec for your database? What happens if you try to check the spec without using Expound e.g. using `s/valid?` or `s/explain` using spec?

We encountered this when using [Guardrails](https://github.com/fulcrologic/guardrails), which in turn leans on expound. We have functions which take a DB as an argument, and we are specing those functions. If a Datomi DB is present as an argument to a function, postwalk _will_ blow up, it doesn't matter if the spec passes or not (it can be `some?`, for example, and the error above still occurs).

We've solved this locally by leaning on Specter rather than `postwalk`. I'll issue a PR so you can look it over.

**Comment by eneroth** [comment]
I'm guessing the same problem would occur with any object that answers `true` to the predicate `map?`, but isn't actually an `IPersistentHashMap`. We just happened upon this with the Datomic DB object.

**Comment by bhb** [comment]
Ah, I see. Thanks for the context and for reporting this bug!

**Comment by bhb** [comment]
@avocade @eneroth Thanks for reporting this bug and for the patch! I ended up going a slightly different direction with the fix, but my patch should still solve the root cause. 

If you're using `deps.edn`, do you mind trying out bdccedc96c65e0d104e6139121a501453007f398 and making sure it solves your issue?

Thanks!

**Comment by avocade** [comment]
Cool! 🙏🏻

**Comment by eneroth** [comment]
> If you're using `deps.edn`, do you mind trying out [bdccedc](https://github.com/bhb/expound/commit/bdccedc96c65e0d104e6139121a501453007f398) and making sure it solves your issue?

I've tested it out, and it all looks green!

**Comment by avocade** [comment]
Please let us know when you have the next release out, so we can ping Tony over at https://github.com/fulcrologic/guardrails with a PR! :)

**Comment by bhb** [comment]
@avocade I've just released 0.8.7 which includes this fix.

**Comment by avocade** [comment]
Cool that's great! Thanks a lot, will try it soon.
